host/linux/posix.c

```diff
@@ -0,0 +1,1133 @@
+        goto done;
+    }
+
+    args->enclaveid = enclaveid;
+    args->epfd = epfd;
+    args->maxevents = (int)maxevents;
+
+    // We lose the wait thread when we exit the func, but the thread will die
+    // on its own copy args then spawn pthread to do the waiting. That way we
+    // can ecall with notification. the thread args are freed by the thread
+    // func.
+    if (pthread_create(&thread, NULL, _epoll_wait_thread, args) < 0)
+    {
+        errno = EINVAL;
+        goto done;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+int oe_posix_epoll_ctl_add_ocall(
+    oe_host_fd_t epfd,
+    oe_host_fd_t fd,
+    unsigned int event_mask,
+    int list_idx,
+    int epoll_enclave_fd)
+{
+    oe_ev_data_t ev_data = {
+        .event_list_idx = (uint32_t)list_idx,
+        .epoll_enclave_fd = (uint32_t)epoll_enclave_fd,
+    };
+    struct epoll_event ev = {
+        .events = event_mask,
+        .data.u64 = ev_data.data,
+    };
+
+    errno = 0;
+
+    return epoll_ctl((int)epfd, EPOLL_CTL_ADD, (int)fd, &ev);
+}
+
+int oe_posix_epoll_ctl_del_ocall(oe_host_fd_t epfd, oe_host_fd_t fd)
+{
+    errno = 0;
+
+    return epoll_ctl((int)epfd, EPOLL_CTL_DEL, (int)fd, NULL);
+}
+
+int oe_posix_epoll_ctl_mod_ocall(
+    oe_host_fd_t epfd,
+    oe_host_fd_t fd,
+    unsigned int event_mask,
+    int list_idx,
+    int enclave_fd)
+{
+    oe_ev_data_t ev_data = {
+        .event_list_idx = (uint32_t)list_idx,
+        .epoll_enclave_fd = (uint32_t)enclave_fd,
+    };
+    struct epoll_event ev = {
+        .events = event_mask,
+        .data.u64 = ev_data.data,
+    };
+
+    return epoll_ctl((int)epfd, EPOLL_CTL_MOD, (int)fd, &ev);
+}
+
+int oe_posix_epoll_close_ocall(oe_host_fd_t fd)
+{
+    errno = 0;
+
+    return close((int)fd);
+}
+
+int oe_posix_shutdown_polling_device_ocall(oe_host_fd_t fd)
+{
+    OE_UNUSED(fd);
+
+    errno = 0;
+
+    return 0;
+}
+
+int oe_posix_epoll_poll_ocall(
+    int64_t enclaveid,
+    oe_host_fd_t epfd,
+    struct oe_pollfd* fds,
+    size_t nfds,
+    int timeout)
+{
+    int ret = -1;
+    size_t fdsize = 0;
+    pthread_t thread = 0;
+    poll_args_t* args = NULL;
+    nfds_t fd_idx = 0;
+
+    errno = 0;
+
+    OE_UNUSED(timeout);
+
+    fdsize = sizeof(struct pollfd) * nfds;
+
+    if (!(args = (poll_args_t*)calloc(1, sizeof(*args) + fdsize)))
+    {
+        errno = ENOMEM;
+        goto done;
+    }
+
+    args->enclaveid = enclaveid;
+    args->epfd = epfd;
+    args->nfds = nfds;
+
+    for (; fd_idx < nfds; fd_idx++)
+    {
+        OE_STATIC_ASSERT(sizeof(args->fds[0]) == sizeof(fds[0]));
+        memcpy(&args->fds[fd_idx], &fds[fd_idx], sizeof(fds[fd_idx]));
+    }
+
+    // We lose the wait thread when we exit the func, but the thread will die
+    // on its own copy args then spawn pthread to do the waiting. That way we
+    // can ecall with notification. the thread args are freed by the thread
+    // func.
+    if (pthread_create(&thread, NULL, _poll_wait_thread, args) < 0)
```

Comment:
@mikbras Shouldn't poll be a blocking call? Right now the poll happens in the created _poll_wait_thread, while the current thread would continue execution and return control back to the enclave. Is the enclave going to spin on some variable till the _poll_wait_thread completes?

@EMumau This seems relevant to switchless calling.
@andschwa Your ideas in this regard also appreciated.

Reply:
Good observation Anand. I brought this up also and I too would like to get rid of this thread. I discussed this with Bruce and it is on my list.

Here is how it works now:

1. The enclave calls **oe_posix_epoll_wait_async_ocall()**, which spins up this thread that watches for events.
2. The enclave returns and then enters **oe_wait_device_notification()** where it waits (on a condition variable) for event notions from (1) the host thread, and (2) other enclave-side devices.

Note that there are at least two event sources: host and enclave.

I proposed an alternative approach.

1. Enclave calls **oe_posix_epoll_wait()** and waits in the host.
2. Control returns to the enclave when events are ready or when another enclave thread wakes the host out of the wait with a function that will be called **oe_posix_epoll_wake()**.

Do you think this is something we should change now, or later?

Reply:
@mikbras 
In current approach, if the enclave is going to wait on a condition variable, the variable in turn would transition back to the host perform a thread wait. Not sure what that is buying us.

The alternative approach is straight-forward and feels more natural.
You can implement the alternative approach as appropriate,  in this PR or the next where the enclave side gets checked in.

Reply:
I cannot defend the original approach since it is not mine and I had the same observation that you did. Therefore: I reworked polling to wait on the host side.